lib/orderbook/MatchingEngine.ts

```diff
@@ -66,34 +101,140 @@ class MatchingEngine {
     };
   }
 
+  /**
+   * Add peer order
+   * @returns false if it's a duplicated order, otherwise true
+   */
+  public addPeerOrder = (order: StampedPeerOrder): boolean => {
+    return this.addOrder(order, this.peerOrders);
+  }
+
   /**
    * Match an order against its opposite queue, and optionally add the unmatched portion of the order to the queue.
    * @param discardRemaining whether to discard any unmatched portion of the order rather than add it to the queue
    * @returns a [[MatchingResult]] with the matches as well as the remaining, unmatched portion of the order
    */
   public matchOrAddOwnOrder = (order: StampedOwnOrder, discardRemaining: boolean): matchingEngine.MatchingResult => {
-    const isBuyOrder = order.quantity > 0;
-    const addTo = isBuyOrder ? this.buyOrders : this.sellOrders;
-
     const matchingResult = this.match(order);
     if (matchingResult.remainingOrder && !discardRemaining) {
-      addTo.add(matchingResult.remainingOrder);
+      this.addOwnOrder(matchingResult.remainingOrder);
     }
 
     return matchingResult;
   }
 
+  private addOwnOrder = (order: StampedOwnOrder): boolean => {
+    return this.addOrder(order, this.ownOrders);
+  }
+
+  private addOrder = (order: StampedOrder, lists: OrderSidesLists<StampedOrder>): boolean => {
+    const isBuyOrder = order.quantity > 0;
+    const list = isBuyOrder ? lists.buy : lists.sell;
+    const queue = isBuyOrder ? this.queues.buy : this.queues.sell;
+
+    if (list.has(order.id)) {
+      return false;
+    } else {
+      list.set(order.id, order);
+      queue.add(order);
+      return true;
+    }
+  }
+
+  public removePeerOrderQuantity = (orderId: string, quantityToDecrease?: number): StampedPeerOrder | undefined => {
```

Comment:
Don't we have a TSLint for having `public` methods before `private` ones? @sangaman